### PR TITLE
Allow gcbmgr and anago to build betas from master

### LINKED ISSUE #1207: Change release branching strategy to create release-1.x branch late
#### Summary

Right now we're creating the release branch at the same date as we cut the `v1.x.0-beta.0` release. After that release we will fast-forward (merge the latest changes from the master branch) periodically into the release branch manually.

The main idea is now to cut the release branch much later, whereas ideally we could get away from the branch fast-forward process.

#### Discussion

Reasons for creating the release branch early:
- It gives extra CI signal (redundant test of same commits on different branches)
- We have extra eyes on code coming from master to release branch (ie: look at all merged commits and affirm they should technically go to the release branch)
- We practice operationally the tooling and processes around shifting content from master to release branch

#### Open Questions
- What would be the right time to cut the release branch? (Code thaw might be too late, whereas code freeze might be sufficient)
- What needs to be done from a release engineering perspective?
- What are the dependencies (some projects might rely on early-cut Kubernetes release branches)?

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind api-change
/kind feature

#### What this PR does / why we need it:
This changes the logic in anago to allow beta builds on the master
branch via a new flag `--type=[alpha|beta|rc|official]`. This flag will
be passed by `krel gcbmgr` and the correctly set GCB substitutions.
#### Which issue(s) this PR fixes:

Fixes #1207

#### Special notes for your reviewer:
I'm not sure about the test-ability of that change. I think the most pragmatic way would be to build the new staging/release images and give it a try in GCB. WDYT?

#### Does this PR introduce a user-facing change?

```release-note
- Added new `anago` flag `--type=[alpha|beta|rc|official]` which deprecates the `--official` and `--rc` flags
- Changed the `krel gcbmgr` `--type` flag to now accept `alpha`, `beta`, `rc` and `official` as well
```

### CODE REVIEW

**Comment by justaugustus** [comment]
FYI @kubernetes/release-engineering @kubernetes/release-managers @onlydole @mrbobbytables @jeremyrickard 
(So everyone gets notifications on this)

**Comment by saschagrunert** [comment]
Cool thanks!

> 1. Should RC numbers start at 0 or 1? We're currently starting RCs at rc.1.

Yes, but I'm wondering if we want to preserve the overall logic while just exchanging `beta` with `rc`. I guess this is this the critical part:

https://github.com/kubernetes/release/blob/1f6ee2f9d35070d3d1d28e5948e340d4af74ddff/lib/releaselib.sh#L420-L433

> 2. Strip/replace all references to `FLAGS_official` and `FLAGS_rc` to cut down on bloat, confusion, and potential side effects. I'm not sure that it's useful to have deadcode around if we have the opportunity to remove it now. Feel free to layer that on as a separate commit, so the intent is clear.

Done, in a separate commit. :heavy_check_mark:

File: anago [hunk 1]
```diff
@@ -838,10 +826,10 @@ copy_staged_from_gcs () {
 # Push git objects to github
 # NOTES:
 # * alpha is alone, pushes tags only
-# * beta is alone, pushes branch and tags
+# * beta is alone, pushes tags only
 # * rc is alone, pushes branch and tags
 # * official pushes both official and beta items - branch and tags
```
Comment (justaugustus) [inline anago:831]:
```suggestion
# * official pushes both official and RC items - branch and tags
```

We want to create a new RC following the official release.

File: anago [hunk 2]
```diff
@@ -838,10 +826,10 @@ copy_staged_from_gcs () {
...
-# * New branch tags a new alpha on master, new beta on new branch and pushes
+# * New branch tags a new alpha or beta on master and pushes
```
Comment (justaugustus) [inline anago:832]:
```suggestion
# * New branch tags a new alpha on master, new RC on new branch, and pushes
```

File: lib/releaselib.sh [hunk 3]
```diff
@@ -418,19 +418,19 @@ release::set_release_version () {
     RELEASE_VERSION[beta]+=".0-beta.0"
     RELEASE_VERSION_PRIME=${RELEASE_VERSION[beta]}
   elif [[ "$parent_branch" =~ release- ]]; then
-    # When we do branched branches we end up with two betas so deal with it
-    # by creating a couple of beta indexes.
-    # beta0 is the branch-point-minor + 1 + beta.1 because
-    # branch-point-minor +1 +beta.0 already exists. This tag lands on the new
+    # When we do branched branches we end up with two rcs so deal with it
```
Comment (justaugustus) [inline lib/releaselib.sh:421]:
```suggestion
    # When we do branched branches we end up with two RCs so deal with it
```

**Comment by saschagrunert** [comment]
> @saschagrunert -- A few nits, but I think it's fine to take these as follow-ups maybe?

Yep, I'll follow-up on the improvements after that PR got mered. :+1:

**Comment by justaugustus** [comment]
Testing...

**Alpha:**

```
$ TOOL_ORG=saschagrunert TOOL_BRANCH=beta-from-master krel gcbmgr --stage --type alpha
```

```
INFO Verifying repository state                   
INFO Repository is in clean state                 
INFO Found matching branch "beta-from-master"     
INFO Found matching organization "saschagrunert" and repository "release" in remote: sascha (https://github.com/saschagrunert/release.git) 
INFO Verifying remote HEAD commit                 
INFO Got remote commit: 3ef6ad2c40165a5ad76cd40e5a600bf1c1dd1e34 
INFO Verifying that remote commit is equal to the local one 
INFO Repository is up-to-date                     
INFO Running gcbmgr with the following options: &{Stage:true Release:false Stream:false Branch:master ReleaseType:alpha BuildVersion: GcpUser: LastJobs:5 Repo:0xc000401930 Version:0xc000401940} 
INFO Build options: {  cloudbuild.yaml   kubernetes-release-test false false false   } 
INFO Retrieving Kubernetes release version for ci/latest on branch master 
INFO Retrieving Kubernetes build version from https://dl.k8s.io/ci/latest.txt... 
INFO Retrieved Kubernetes version: v1.19.0-alpha.3.252+438debfb726361 
INFO Trying to get the kube-cross version for master... 
INFO Found the following kube-cross version: v1.13.9-5 
INFO Listing GCB substitutions prior to build submission... 
INFO TYPE: alpha                                  
INFO BUILDVERSION: --buildversion=v1.19.0-alpha.3.252+438debfb726361 
INFO TOOL_ORG: saschagrunert                      
INFO TOOL_REPO: release                           
INFO GCP_USER_TAG: stephen-k8s-at-agst-us         
INFO BUILD_AT_HEAD:                               
INFO BUILD_POINT: v1.19.0-alpha.3.252-438debfb726361 
INFO KUBE_CROSS_VERSION: v1.13.9-5                
INFO NOMOCK_TAG:                                  
INFO NOMOCK:                                      
INFO TOOL_BRANCH: beta-from-master                
INFO TYPE_TAG: alpha                              
INFO RELEASE_BRANCH: master                       
INFO Build directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
INFO Config directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
INFO cd-ing to build directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
ID                                    CREATE_TIME                DURATION  SOURCE  IMAGES  STATUS
05b642f8-8802-449f-915a-1245651044bf  2020-05-14T14:49:13+00:00  -         -       -       QUEUED
INFO: Refreshing access_token
Created [https://cloudbuild.googleapis.com/v1/projects/kubernetes-release-test/builds/05b642f8-8802-449f-915a-1245651044bf].
Logs are available at [https://console.cloud.google.com/cloud-build/builds/05b642f8-8802-449f-915a-1245651044bf?project=648026197307].
INFO: Display format: "
        table(
          id,
          createTime.date('%Y-%m-%dT%H:%M:%S%Oz', undefined='-'),
          duration(start=startTime,end=finishTime,precision=0,calendar=false,undefined="  -").slice(2:).join(""):label=DURATION,
          build_source(undefined="-"):label=SOURCE,
          build_images(undefined="-"):label=IMAGES,
          status
        )
      "
```

**Beta:**

```
$ TOOL_ORG=saschagrunert TOOL_BRANCH=beta-from-master krel gcbmgr --stage --type beta
```

```
INFO Verifying repository state                   
INFO Repository is in clean state                 
INFO Found matching branch "beta-from-master"     
INFO Found matching organization "saschagrunert" and repository "release" in remote: sascha (https://github.com/saschagrunert/release.git) 
INFO Verifying remote HEAD commit                 
INFO Got remote commit: 3ef6ad2c40165a5ad76cd40e5a600bf1c1dd1e34 
INFO Verifying that remote commit is equal to the local one 
INFO Repository is up-to-date                     
INFO Running gcbmgr with the following options: &{Stage:true Release:false Stream:false Branch:master ReleaseType:beta BuildVersion: GcpUser: LastJobs:5 Repo:0xc000411920 Version:0xc000411930} 
INFO Build options: {  cloudbuild.yaml   kubernetes-release-test false false false   } 
INFO Retrieving Kubernetes release version for ci/latest on branch master 
INFO Retrieving Kubernetes build version from https://dl.k8s.io/ci/latest.txt... 
INFO Retrieved Kubernetes version: v1.19.0-alpha.3.252+438debfb726361 
INFO Trying to get the kube-cross version for master... 
INFO Found the following kube-cross version: v1.13.9-5 
INFO Listing GCB substitutions prior to build submission... 
INFO GCP_USER_TAG: stephen-k8s-at-agst-us         
INFO TYPE: beta                                   
INFO TYPE_TAG: beta                               
INFO RELEASE_BRANCH: master                       
INFO BUILD_AT_HEAD:                               
INFO BUILDVERSION: --buildversion=v1.19.0-alpha.3.252+438debfb726361 
INFO TOOL_BRANCH: beta-from-master                
INFO TOOL_REPO: release                           
INFO BUILD_POINT: v1.19.0-alpha.3.252-438debfb726361 
INFO KUBE_CROSS_VERSION: v1.13.9-5                
INFO NOMOCK_TAG:                                  
INFO NOMOCK:                                      
INFO TOOL_ORG: saschagrunert                      
INFO Build directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
INFO Config directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
INFO cd-ing to build directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
ID                                    CREATE_TIME                DURATION  SOURCE  IMAGES  STATUS
c8735edf-8363-4140-92e3-2a6564a0f7a5  2020-05-14T14:49:45+00:00  -         -       -       QUEUED
Created [https://cloudbuild.googleapis.com/v1/projects/kubernetes-release-test/builds/c8735edf-8363-4140-92e3-2a6564a0f7a5].
Logs are available at [https://console.cloud.google.com/cloud-build/builds/c8735edf-8363-4140-92e3-2a6564a0f7a5?project=648026197307].
INFO: Display format: "
        table(
          id,
          createTime.date('%Y-%m-%dT%H:%M:%S%Oz', undefined='-'),
          duration(start=startTime,end=finishTime,precision=0,calendar=false,undefined="  -").slice(2:).join(""):label=DURATION,
          build_source(undefined="-"):label=SOURCE,
          build_images(undefined="-"):label=IMAGES,
          status
        )
      "
```

**RC:**

```
$ TOOL_ORG=saschagrunert TOOL_BRANCH=beta-from-master krel gcbmgr --stage --type rc --branch release-1.19
```

```
INFO Verifying repository state                   
INFO Repository is in clean state                 
INFO Found matching branch "beta-from-master"     
INFO Found matching organization "saschagrunert" and repository "release" in remote: sascha (https://github.com/saschagrunert/release.git) 
INFO Verifying remote HEAD commit                 
INFO Got remote commit: 3ef6ad2c40165a5ad76cd40e5a600bf1c1dd1e34 
INFO Verifying that remote commit is equal to the local one 
INFO Repository is up-to-date                     
INFO Running gcbmgr with the following options: &{Stage:true Release:false Stream:false Branch:release-1.19 ReleaseType:rc BuildVersion: GcpUser: LastJobs:5 Repo:0xc00040f920 Version:0xc00040f930} 
INFO Build options: {  cloudbuild.yaml   kubernetes-release-test false false false   } 
INFO Retrieving Kubernetes release version for ci/latest on branch release-1.19 
INFO Retrieving Kubernetes build version from https://dl.k8s.io/ci/latest-1.19.txt... 
INFO Retrieved Kubernetes version: v1.19.0-alpha.3.252+438debfb726361 
INFO Trying to get the kube-cross version for release-1.19... 
INFO Error retrieving the kube-cross version for the 'release-1.19': HTTP status not OK (404) for https://raw.githubusercontent.com/kubernetes/kubernetes/release-1.19/build/build-image/cross/VERSION 
INFO Trying to get the kube-cross version for master... 
INFO Found the following kube-cross version: v1.13.9-5 
INFO Listing GCB substitutions prior to build submission... 
INFO TOOL_REPO: release                           
INFO TOOL_BRANCH: beta-from-master                
INFO RELEASE_BRANCH: release-1.19                 
INFO BUILD_POINT: v1.19.0-alpha.3.252-438debfb726361 
INFO KUBE_CROSS_VERSION: v1.13.9-5                
INFO BUILDVERSION: --buildversion=v1.19.0-alpha.3.252+438debfb726361 
INFO NOMOCK_TAG:                                  
INFO NOMOCK:                                      
INFO TOOL_ORG: saschagrunert                      
INFO GCP_USER_TAG: stephen-k8s-at-agst-us         
INFO TYPE: rc                                     
INFO TYPE_TAG: rc                                 
INFO BUILD_AT_HEAD:                               
INFO Build directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
INFO Config directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
INFO cd-ing to build directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
ID                                    CREATE_TIME                DURATION  SOURCE  IMAGES  STATUS
47fc8563-7734-4ec4-87e9-2d2c2d370139  2020-05-14T14:50:01+00:00  -         -       -       QUEUED
Created [https://cloudbuild.googleapis.com/v1/projects/kubernetes-release-test/builds/47fc8563-7734-4ec4-87e9-2d2c2d370139].
Logs are available at [https://console.cloud.google.com/cloud-build/builds/47fc8563-7734-4ec4-87e9-2d2c2d370139?project=648026197307].
INFO: Display format: "
        table(
          id,
          createTime.date('%Y-%m-%dT%H:%M:%S%Oz', undefined='-'),
          duration(start=startTime,end=finishTime,precision=0,calendar=false,undefined="  -").slice(2:).join(""):label=DURATION,
          build_source(undefined="-"):label=SOURCE,
          build_images(undefined="-"):label=IMAGES,
          status
        )
      "
```

**Official:**

```
$ TOOL_ORG=saschagrunert TOOL_BRANCH=beta-from-master krel gcbmgr --stage --type official --branch release-1.19
```

```
INFO Verifying repository state                   
INFO Repository is in clean state                 
INFO Found matching branch "beta-from-master"     
INFO Found matching organization "saschagrunert" and repository "release" in remote: sascha (https://github.com/saschagrunert/release.git) 
INFO Verifying remote HEAD commit                 
INFO Got remote commit: 3ef6ad2c40165a5ad76cd40e5a600bf1c1dd1e34 
INFO Verifying that remote commit is equal to the local one 
INFO Repository is up-to-date                     
INFO Running gcbmgr with the following options: &{Stage:true Release:false Stream:false Branch:release-1.19 ReleaseType:official BuildVersion: GcpUser: LastJobs:5 Repo:0xc000403930 Version:0xc000403940} 
INFO Build options: {  cloudbuild.yaml   kubernetes-release-test false false false   } 
INFO Retrieving Kubernetes release version for ci/latest on branch release-1.19 
INFO Retrieving Kubernetes build version from https://dl.k8s.io/ci/latest-1.19.txt... 
INFO Retrieved Kubernetes version: v1.19.0-alpha.3.252+438debfb726361 
INFO Trying to get the kube-cross version for release-1.19... 
INFO Error retrieving the kube-cross version for the 'release-1.19': HTTP status not OK (404) for https://raw.githubusercontent.com/kubernetes/kubernetes/release-1.19/build/build-image/cross/VERSION 
INFO Trying to get the kube-cross version for master... 
INFO Found the following kube-cross version: v1.13.9-5 
INFO Listing GCB substitutions prior to build submission... 
INFO BUILD_AT_HEAD:                               
INFO BUILD_POINT: v1.19.0-alpha.3.252-438debfb726361 
INFO KUBE_CROSS_VERSION: v1.13.9-5                
INFO TOOL_ORG: saschagrunert                      
INFO TOOL_REPO: release                           
INFO TOOL_BRANCH: beta-from-master                
INFO RELEASE_BRANCH: release-1.19                 
INFO BUILDVERSION: --buildversion=v1.19.0-alpha.3.252+438debfb726361 
INFO NOMOCK_TAG:                                  
INFO NOMOCK:                                      
INFO GCP_USER_TAG: stephen-k8s-at-agst-us         
INFO TYPE: official                               
INFO TYPE_TAG: official                           
INFO Build directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
INFO Config directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
INFO cd-ing to build directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
ID                                    CREATE_TIME                DURATION  SOURCE  IMAGES  STATUS
b28b0cbf-2d75-4fe5-9c52-d7b46a9abc68  2020-05-14T14:50:34+00:00  -         -       -       QUEUED
Created [https://cloudbuild.googleapis.com/v1/projects/kubernetes-release-test/builds/b28b0cbf-2d75-4fe5-9c52-d7b46a9abc68].
Logs are available at [https://console.cloud.google.com/cloud-build/builds/b28b0cbf-2d75-4fe5-9c52-d7b46a9abc68?project=648026197307].
INFO: Display format: "
        table(
          id,
          createTime.date('%Y-%m-%dT%H:%M:%S%Oz', undefined='-'),
          duration(start=startTime,end=finishTime,precision=0,calendar=false,undefined="  -").slice(2:).join(""):label=DURATION,
          build_source(undefined="-"):label=SOURCE,
          build_images(undefined="-"):label=IMAGES,
          status
        )
      "
```

**Comment by justaugustus** [comment]
/hold cancel
Merging this so I can work some fixes on top of it.

**Pull request merged**